the-collab-lab/tcl-7-smart-shopping-list
# Adding a Filter List Feature

## Description
This Pull Request adds a list filtering functionality to the user's grocery list.  We've add an input field that then, based on what the use types, filters through their grocery list and displays only the items that match the input text.

As we're refactoring and changing class components to functional ones, we used this tutorial to guide us in creating the filtering functionality. [https://dev.to/asimdahall/simple-search-form-in-react-using-hooks-42pg](https://dev.to/asimdahall/simple-search-form-in-react-using-hooks-42pg)

We were warned to be aware and careful developing this functionality as in previous cohorts, the filtering capability sends a ton of calls to the database.  
- Our first approach was to create the filter function as a separate component and then import that component into the `getList` component but we were running into issues connecting the components.
- Our second approach was to add the component as a function in the `getList` component but had some issues getting the `props.data` values to pass to the function
- Our successful approach brought the functionality into the `<FullList />` component that gave our filtering function access to the data props.

Issues we ran into:
- Adding the filtering function to the `props.data.map` function triggered the "tons of database calls" we were warned about.  We then created a new variable to store the results of the filter function and rendered either the filtered grocery list or the whole grocery list.  This still looks like it's making lots of database calls but definitely seems to be quite a bit fewer in number.
- We couldn't figure out how to apply the checkbox input to both lists without write redundant code.  We tried creating a variable to store the data props but couldn't get and connect the properties need for the checkbox input, they show up as undefined in the console.

 ## Todo
- [x] We'd like to refactor this to store the data props in a variable that can be used by different functions.

## Related Issue

closes #10  

## Acceptance Criteria

- [x] Display a text field at the top of the shopping list
- [x] As the user types into the field, the list narrows to include only items that contain the text the user entered in the filter field

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

![grocery-before](https://user-images.githubusercontent.com/2355543/83926937-25ac0f00-a740-11ea-9602-f0acb5f31e13.png)



### After

![grocery-after](https://user-images.githubusercontent.com/2355543/83926947-2d6bb380-a740-11ea-810e-fd0036d371f0.png)



## Testing Steps / QA Criteria

From your terminal, pull down this branch with git pull origin `dt-mm-list-filter-feature` and check that branch out with git checkout `dt-mm-list-filter-feature`

Then npm install to install the newly added dependencies locally and npm start to launch the app.

Navigate to the `list-view` page of the app.  In the Filter List field, start typing the name of one of the grocery list items.  The list should shrink to only show the item you've entered.  To reset and view the whole list, press the x button and the filter input text should clear and all of the grocery list items should be displayed.

Need some advice on accessibility of the filter list input field, it's not passing even though it's wrapped in a label and has placeholder text.
